### ISSUE TITLE
What is debugging like?

### ISSUE THREAD

**Issue by jedwards1211** [dwyl/learn-elm#6]
@nelsonic I know you said there are really good stack traces, but I'm also curious: how are the source maps?  Is it easy to put a breakpoint in the Elm code and step through it in Chrome dev tools?  I guess this would be my main concern with a non-JS based language.  I've found it easy enough to read and set breakpoints in Babel-transpiled ES6 at least (since using full sourcemaps for production-scale apps is really slow on each recompile).

**Comment by nelsonic** [comment]
@jedwards1211 _really_ good question! (_thank you for asking it!_)

**Comment by jedwards1211** [comment]
@nelsonic yeah if I remember correctly that was the inspiration for the redux dev tools, which do essentially the same thing, though in my experience with a complex app with a high volume of events, it was quite slow.

And the time traveling debugger is great but it's only half of the battle; what's missing from that demo is putting breakpoints in the code and stepping through to see what went wrong when some action was mishandled.

**Comment by jedwards1211** [comment]
@nelsonic strangely Evan Czaplicki [claims](https://github.com/elm-lang/elm-compiler/issues/1273#issuecomment-183518010) that few companies in production have much of a need for source maps so they aren't a priority right now.  I don't yet understand what about Elm enables them to avoid using any source maps.  I asked him about it in that thread.

**Comment by jedwards1211** [comment]
Googling for "elm breakpoints", I didn't find much.  But I found a few posts on the google groups for Elm.  One user [summarizes](https://groups.google.com/d/msg/elm-discuss/mwFQOp8dwh0/sA8KZF2wBgAJ) my biggest reservation about using Elm in production-scale web apps:

> Elm, as great as it is, doesn't save you from debugging every once in a while. The one option we have now, is logging. Logging is great, but it can quickly become painful in loops. Since Elm compiles to a single JS file with long mangled names, setting a breakpoint from the code would sometimes be the simplest way to properly debug your code. The generated JS isn't that hard to understand either, it's a series of vars and function calls for the most part.

**Comment by nelsonic** [comment]
@jedwards1211 _thank you **so much**_ for digging into the debugging "_story_"... ❤️  ✅

**Comment by SimonLab** [comment]
It seems that more debugging tools will be available also on the next version of elm (0.18): https://youtu.be/DSjbTC-hvqQ?list=PLglJM3BYAMPH2zuz1nbKHQyeawE4SN0Cd&t=1633

**Comment by nelsonic** [comment]
@SimonLab yeah, `v0.18` looks _very_ promising from a Debugging (& "QA") perspective. 😉 
I _think_ @jedwards1211 was more focussed on the ability to use "break points" in the code and step through it: https://developers.google.com/web/tools/chrome-devtools/javascript/add-breakpoints
But in my (_probably naive_) view writing smaller ("_pure_") functions and having a (_really_) smart compiler that looks for typos and type errors will go a _long_ way toward improving the debugging story ... while this does not directly answer the question @jedwards1211 _originally_ asked - _and I think we should `continue` to ask it_ - I'm still _optimistic_ ... 🌈

**Comment by nelsonic** [comment]
@jedwards1211 this _might_ interest you: http://elm-lang.org/blog/the-perfect-bug-report 😉

**Comment by jedwards1211** [comment]
Yes, importing/exporting history is very helpful and fortunately possible with Redux as well.

**Comment by nelsonic** [comment]
@jedwards1211, _indeed_, Dan Abramov did a _great_ job of "_porting_" the elm debugger over to Redux. 😉
